File: 047.rb
# 047
# Permutations II
# 2015-09-04
#####################################################
# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def next_permutation(nums)
  n = nums.size
  return false if n <= 1
  i = n - 2
  while i >= 0 and nums[i] >= nums[i+1]
    i-=1
  end
  if i >= 0
    index = nums.each_with_index.find_all{|value, index| index > i and value > nums[i]}.min[1]
    nums[i], nums[index] = nums[index], nums[i]
  end
  nums[i+1..-1] = nums[i+1..-1].sort
  return i >= 0
end

# @param {Integer[]} nums
# @return {Integer[][]}
def permute_unique(nums)
  nums.sort!
  v = [nums]
  while next_permutation(nums)
    v.push(nums.dup)
  end
  v
end

#####################################################

nums = [1,2,3]
v = permute_unique(nums)
puts v.inspect

File: python/047.py
#   047
#   Permutations II
#   2015-09-04
#####################################################
class Solution:
    # @param {integer[]} nums
    # @return {void} Do not return anything, modify nums in-place instead.
    def nextPermutation(self, nums):
        n = len(nums)
        if n <= 1:
            return False
        i = n - 2
        while i >= 0 and nums[i] >= nums[i+1]:
            i -= 1
        if i >= 0:
            b = min([a for a in nums[i+1:] if a > nums[i]])
            index = i+1 + nums[i+1:].index(b)
            nums[i], nums[index] = nums[index], nums[i]
        nums[i+1:] = sorted(nums[i+1:])
        return i >= 0

    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums = sorted(nums)
        v = [nums[:]]
        while self.nextPermutation(nums):
            v.append(nums[:])
        return v
        
#####################################################

nums = [1,3,2]
sol = Solution()
z = sol.permuteUnique(nums)
print(z)

